File: src/services/obd2/gearCalculator.ts
import { CIVIC_2013_SPECS } from './civicSpecs';

export type GearSelection = 1 | 2 | 3 | 4 | 5 | 'N' | 'CLUTCH';

export interface GearAnalysisResult {
  currentGear: GearSelection;
  calculatedRatio: number;
  expectedRatio: number;
  ratioToleranceDelta: number;
  isClutchSlipping: boolean;
  optimalShiftRpm: number;
  shouldShiftUp: boolean;
  shiftLightStage: number; // 0 to 5 (0: normal, 1-3: approaching, 4: optimal shift, 5: flashing redline)
}

export class GearCalculatorEngine {
  private previousRpm: number = 0;
  private previousSpeedKmh: number = 0;
  private previousTimestamp: number = Date.now();
  private slipConfirmationCounter: number = 0;

  // Expected overall transmission gear ratio (RPM / Wheel RPM)
  // Wheel RPM = (SpeedKmh / 60) / TireCircumferenceKm
  // Total Ratio = Engine RPM / Wheel RPM = Gear Ratio * Final Drive Ratio
  private targetOverallRatios: Record<1 | 2 | 3 | 4 | 5, number>;

  constructor() {
    this.targetOverallRatios = {
      1: CIVIC_2013_SPECS.gearRatios[1] * CIVIC_2013_SPECS.finalDriveRatio, // 3.143 * 4.294 = 13.496
      2: CIVIC_2013_SPECS.gearRatios[2] * CIVIC_2013_SPECS.finalDriveRatio, // 1.870 * 4.294 = 8.030
      3: CIVIC_2013_SPECS.gearRatios[3] * CIVIC_2013_SPECS.finalDriveRatio, // 1.235 * 4.294 = 5.303
      4: CIVIC_2013_SPECS.gearRatios[4] * CIVIC_2013_SPECS.finalDriveRatio, // 0.949 * 4.294 = 4.075
      5: CIVIC_2013_SPECS.gearRatios[5] * CIVIC_2013_SPECS.finalDriveRatio, // 0.727 * 4.294 = 3.122
    };
  }

  /**
   * Evaluates the active gear based on real-time RPM and vehicle speed.
   */
  public analyzeGear(
    rpm: number,
    speedKmh: number,
    throttlePercent: number,
    shiftMode: 'eco' | 'power' = 'eco'
  ): GearAnalysisResult {
    const now = Date.now();
    const dt = Math.max(0.05, (now - this.previousTimestamp) / 1000);

    // If car is stationary or barely moving (< 3 km/h)
    if (speedKmh < 3.0) {
      this.previousRpm = rpm;
      this.previousSpeedKmh = speedKmh;
      this.previousTimestamp = now;
      return {
        currentGear: rpm > 1100 ? 'CLUTCH' : 'N',
        calculatedRatio: 0,
        expectedRatio: 0,
        ratioToleranceDelta: 0,
        isClutchSlipping: false,
        optimalShiftRpm: CIVIC_2013_SPECS.ecoShiftPoints[1],
        shouldShiftUp: false,
        shiftLightStage: 0,
      };
    }

    // Calculate Wheel RPM from Speed and Civic Tire Circumference
    const wheelRpm = (speedKmh / 60) / CIVIC_2013_SPECS.tireCircumferenceKm;
    const currentOverallRatio = wheelRpm > 0 ? rpm / wheelRpm : 0;

    // Match with 5-speed gear ratios within a ±7% tolerance window
    let detectedGear: GearSelection = 'N';
    let bestDelta = 999;
    let expectedRatio = 0;

    const gears: (1 | 2 | 3 | 4 | 5)[] = [1, 2, 3, 4, 5];

    for (const g of gears) {
      const target = this.targetOverallRatios[g];
      const deltaPercent = Math.abs(currentOverallRatio - target) / target;

      if (deltaPercent <= 0.08 && deltaPercent < bestDelta) {
        bestDelta = deltaPercent;
        detectedGear = g;
        expectedRatio = target;
      }
    }

    // If ratio doesn't match any gear:
    // If RPM is significantly higher than expected (e.g. revving with clutch pressed)
    // or RPM dropped to idle (750-900) while rolling fast -> 'N' or 'CLUTCH'
    if (detectedGear === 'N') {
      if (rpm <= 1000 && speedKmh > 10) {
        detectedGear = 'N'; // Rolling in neutral with engine at idle
      } else {
        detectedGear = 'CLUTCH'; // Shifting / clutch disengaged / rev-matching
      }
    }

    // Clutch Slip Detection:
    // If vehicle is in gear, throttle is high (> 35%), RPM is rising rapidly (> 1200 RPM/s)
    // but vehicle speed is NOT accelerating proportionally:
    let isClutchSlipping = false;
    if (typeof detectedGear === 'number' && throttlePercent > 35 && rpm > 2500) {
      const rpmRate = (rpm - this.previousRpm) / dt;
      const speedRate = (speedKmh - this.previousSpeedKmh) / dt;

      // In gear, RPM rate should directly match speed rate
      // If RPM is flaring upwards with near-zero vehicle acceleration:
      if (rpmRate > 1200 && speedRate < 1.0) {
        this.slipConfirmationCounter++;
        if (this.slipConfirmationCounter >= 3) {
          isClutchSlipping = true;
        }
      } else {
        this.slipConfirmationCounter = Math.max(0, this.slipConfirmationCounter - 1);
      }
    } else {
      this.slipConfirmationCounter = 0;
    }

    // Shift Light & Shift Point calculation
    let optimalShiftRpm = 6500;
    let shouldShiftUp = false;
    let shiftLightStage = 0;

    if (typeof detectedGear === 'number') {
      if (shiftMode === 'eco') {
        optimalShiftRpm = CIVIC_2013_SPECS.ecoShiftPoints[detectedGear as 1 | 2 | 3 | 4] || 2500;
        
        if (detectedGear < 5 && rpm >= optimalShiftRpm) {
          shouldShiftUp = true;
        }
        
        // Eco shift stages (0-5)
        if (rpm < 1800) shiftLightStage = 0;
        else if (rpm < 2000) shiftLightStage = 1;
        else if (rpm < 2150) shiftLightStage = 2;
        else if (rpm < optimalShiftRpm + 100) shiftLightStage = 3;
        else if (rpm < 3000) shiftLightStage = 4; // Shift now indicator
        else shiftLightStage = 5; // Exceeded eco threshold
      } else {
        // Power / Sport Shift Light
        optimalShiftRpm = CIVIC_2013_SPECS.powerShiftPointRpm; // 6,500 RPM
        shouldShiftUp = detectedGear < 5 && rpm >= 6300;

        if (rpm < 3500) shiftLightStage = 0;
        else if (rpm < 4500) shiftLightStage = 1; // Green 1
        else if (rpm < 5200) shiftLightStage = 2; // Green 2 / VTEC window
        else if (rpm < 5900) shiftLightStage = 3; // Yellow
        else if (rpm < 6400) shiftLightStage = 4; // Orange / Peak HP
        else shiftLightStage = 5; // Flashing Red Redline
      }
    }

    this.previousRpm = rpm;
    this.previousSpeedKmh = speedKmh;
    this.previousTimestamp = now;

    return {
      currentGear: detectedGear,
      calculatedRatio: currentOverallRatio,
      expectedRatio,
      ratioToleranceDelta: bestDelta !== 999 ? bestDelta : 0,
      isClutchSlipping,
      optimalShiftRpm,
      shouldShiftUp,
      shiftLightStage,
    };
  }
}

File: app-native/core/src/main/kotlin/com/shieldrj/civic5mt/core/GearCalculator.kt
package com.shieldrj.civic5mt.core

import kotlin.math.abs
import kotlin.math.max

/**
 * Which gear the car is in, as far as this app can tell.
 *
 * A sealed type rather than the TypeScript's `1 | 2 | 3 | 4 | 5 | 'N' | 'CLUTCH'`, which
 * mixed numbers and strings in one value and forced a `typeof` check at every use. The
 * distinction that matters - "in a numbered gear" versus "not" - is now something the
 * compiler enforces instead of something each caller remembers to test for.
 */
sealed interface GearSelection {
    data class Gear(val number: Int) : GearSelection {
        override fun toString(): String = number.toString()
    }

    /** Rolling or stopped with the engine disconnected from the wheels. */
    data object Neutral : GearSelection {
        override fun toString(): String = "N"
    }

    /** Clutch in: shifting, rev-matching, or revving while stationary. */
    data object Clutch : GearSelection {
        override fun toString(): String = "CLUTCH"
    }
}

enum class ShiftMode { ECO, POWER }

data class GearAnalysisResult(
    val currentGear: GearSelection,
    val calculatedRatio: Double,
    val expectedRatio: Double,
    val ratioToleranceDelta: Double,
    val isClutchSlipping: Boolean,
    val optimalShiftRpm: Int,
    val shouldShiftUp: Boolean,
    /** 0 to 5 (0: normal, 1-3: approaching, 4: optimal shift, 5: flashing redline). */
    val shiftLightStage: Int,
)

class GearCalculatorEngine(private val clock: MillisClock = SystemMillisClock) {

    private var previousRpm: Double = 0.0
    private var previousSpeedKmh: Double = 0.0
    private var previousTimestamp: Long = clock.nowMillis()
    private var slipConfirmationCounter: Int = 0

    /**
     * Expected overall transmission gear ratio (engine RPM / wheel RPM).
     * Wheel RPM = (speed km/h / 60) / tyre circumference km.
     * Total ratio = gear ratio * final drive ratio.
     */
    private val targetOverallRatios: Map<Int, Double> =
        CivicSpecs.GEAR_RATIOS.mapValues { (_, ratio) -> ratio * CivicSpecs.FINAL_DRIVE_RATIO }

    /**
     * Evaluates the active gear based on real-time RPM and vehicle speed.
     */
    fun analyzeGear(
        rpm: Double,
        speedKmh: Double,
        throttlePercent: Double,
        shiftMode: ShiftMode = ShiftMode.ECO,
    ): GearAnalysisResult {
        val now = clock.nowMillis()
        val dt = max(0.05, (now - previousTimestamp) / 1000.0)

        // Stationary or barely moving (< 3 km/h).
        if (speedKmh < 3.0) {
            previousRpm = rpm
            previousSpeedKmh = speedKmh
            previousTimestamp = now
            return GearAnalysisResult(
                currentGear = if (rpm > 1100) GearSelection.Clutch else GearSelection.Neutral,
                calculatedRatio = 0.0,
                expectedRatio = 0.0,
                ratioToleranceDelta = 0.0,
                isClutchSlipping = false,
                optimalShiftRpm = CivicSpecs.ECO_SHIFT_POINTS.getValue(1),
                shouldShiftUp = false,
                shiftLightStage = 0,
            )
        }

        // Wheel RPM from speed and the Civic's tyre circumference.
        val wheelRpm = (speedKmh / 60.0) / CivicSpecs.TIRE_CIRCUMFERENCE_KM
        val currentOverallRatio = if (wheelRpm > 0) rpm / wheelRpm else 0.0

        // Match against the 5-speed ratios within an 8% tolerance window.
        var detectedGearNumber: Int? = null
        var bestDelta = 999.0
        var expectedRatio = 0.0

        for (g in 1..5) {
            val target = targetOverallRatios.getValue(g)
            val deltaPercent = abs(currentOverallRatio - target) / target

            if (deltaPercent <= 0.08 && deltaPercent < bestDelta) {
                bestDelta = deltaPercent
                detectedGearNumber = g
                expectedRatio = target
            }
        }

        // If the ratio matched no gear: idling while rolling is neutral, anything else is
        // the clutch being in - shifting, rev-matching, or a disengaged drivetrain.
        val detectedGear: GearSelection = when {
            detectedGearNumber != null -> GearSelection.Gear(detectedGearNumber)
            rpm <= 1000 && speedKmh > 10 -> GearSelection.Neutral
            else -> GearSelection.Clutch
        }

        // Clutch slip: in gear, throttle open, RPM flaring, but the car is not accelerating
        // with it. Confirmed over three consecutive ticks so one noisy sample cannot raise it.
        var isClutchSlipping = false
        if (detectedGear is GearSelection.Gear && throttlePercent > 35 && rpm > 2500) {
            val rpmRate = (rpm - previousRpm) / dt
            val speedRate = (speedKmh - previousSpeedKmh) / dt

            if (rpmRate > 1200 && speedRate < 1.0) {
                slipConfirmationCounter++
                if (slipConfirmationCounter >= 3) {
                    isClutchSlipping = true
                }
            } else {
                slipConfirmationCounter = max(0, slipConfirmationCounter - 1)
            }
        } else {
            slipConfirmationCounter = 0
        }

        // Shift light and shift point.
        var optimalShiftRpm = 6500
        var shouldShiftUp = false
        var shiftLightStage = 0

        if (detectedGear is GearSelection.Gear) {
            val gear = detectedGear.number
            if (shiftMode == ShiftMode.ECO) {
                optimalShiftRpm = CivicSpecs.ECO_SHIFT_POINTS[gear] ?: 2500

                if (gear < 5 && rpm >= optimalShiftRpm) {
                    shouldShiftUp = true
                }

                shiftLightStage = when {
                    rpm < 1800 -> 0
                    rpm < 2000 -> 1
                    rpm < 2150 -> 2
                    rpm < optimalShiftRpm + 100 -> 3
                    rpm < 3000 -> 4 // Shift now
                    else -> 5 // Past the eco threshold
                }
            } else {
                optimalShiftRpm = CivicSpecs.POWER_SHIFT_POINT_RPM
                shouldShiftUp = gear < 5 && rpm >= 6300

                shiftLightStage = when {
                    rpm < 3500 -> 0
                    rpm < 4500 -> 1 // Green 1
                    rpm < 5200 -> 2 // Green 2 / VTEC window
                    rpm < 5900 -> 3 // Yellow
                    rpm < 6400 -> 4 // Orange / peak HP
                    else -> 5 // Flashing red, redline
                }
            }
        }

        previousRpm = rpm
        previousSpeedKmh = speedKmh
        previousTimestamp = now

        return GearAnalysisResult(
            currentGear = detectedGear,
            calculatedRatio = currentOverallRatio,
            expectedRatio = expectedRatio,
            ratioToleranceDelta = if (bestDelta != 999.0) bestDelta else 0.0,
            isClutchSlipping = isClutchSlipping,
            optimalShiftRpm = optimalShiftRpm,
            shouldShiftUp = shouldShiftUp,
            shiftLightStage = shiftLightStage,
        )
    }
}
